singer_sdk/contrib/filesystem/stream.py

```diff
@@ -0,0 +1,141 @@
+"""Stream class for reading from files."""
+
+from __future__ import annotations
+
+import abc
+import functools
+import typing as t
+
+from singer_sdk import Stream
+from singer_sdk.exceptions import ConfigValidationError
+from singer_sdk.helpers._util import utc_now
+from singer_sdk.streams.core import REPLICATION_INCREMENTAL
+
+if t.TYPE_CHECKING:
+    import datetime
+
+    import fsspec
+
+    from singer_sdk.helpers.types import Context, Record
+    from singer_sdk.tap_base import Tap
+
+SDC_META_FILEPATH = "_sdc_path"
+SDC_META_MODIFIED_AT = "_sdc_modified_at"
+
+
+class FileStream(Stream, metaclass=abc.ABCMeta):
+    """Abstract base class for file streams."""
+
+    SDC_PROPERTIES: t.ClassVar[dict[str, dict]] = {
+        SDC_META_FILEPATH: {"type": "string"},
+        SDC_META_MODIFIED_AT: {"type": ["string", "null"], "format": "date-time"},
+    }
+
+    def __init__(
+        self,
+        tap: Tap,
+        name: str,
+        *,
+        filepaths: t.Sequence[str],
+        filesystem: fsspec.AbstractFileSystem,
+    ) -> None:
+        """Create a new FileStream instance.
+
+        Args:
+            tap: The tap for this stream.
+            name: The name of the stream.
+            filepaths: List of file paths to read.
+            filesystem: The filesystem implementation object to use.
+            mode: The read mode for the stream.
+
+        Raises:
+            ConfigValidationError: If no file paths are provided.
+        """
+        if not filepaths:  # pragma: no cover
+            msg = "Configuration error"
+            raise ConfigValidationError(msg, errors=["No file paths provided"])
+
+        self._filepaths = filepaths
+        self.filesystem = filesystem
+
+        super().__init__(tap, schema=None, name=name)
+
+        # TODO(edgarrmondragon): Make this None if the filesytem does not support it.
+        self.replication_key = SDC_META_MODIFIED_AT
+        self._sync_start_time = utc_now()
+        self._partitions = [{SDC_META_FILEPATH: path} for path in self._filepaths]
+
+    @property
+    def partitions(self) -> list[dict[str, t.Any]]:
+        """Return the list of partitions for this stream."""
+        return self._partitions
```

Comment:
Using partitions allows us to track state for each individual file in _merge_ mode.